drivers/sdhc/rcar_mmc.c

```diff
@@ -452,11 +471,77 @@ static uint32_t rcar_mmc_gen_data_cmd(struct sdhc_command *cmd,
 	return cmd_reg;
 }
 
+/**
+ * @brief Transmit/Receive data to/from MMC using DMA
+ *
+ * Sends/Receives data to/from the MMC controller.
+ *
+ * @note in/out parameters should be checked by a caller function.
+ *
+ * @param dev MMC device
+ * @param data MMC data buffer for tx/rx
+ * @param is_read it is read or write operation
+ *
+ * @retval 0 tx/rx was successful
+ * @retval -ENOTSUP: cache flush/invalidate aren't supported
+ * @retval -ETIMEDOUT: timed out while tx/rx
+ * @retval -EIO: I/O error
+ * @retval -EILSEQ: communication out of sync
+ */
 static int rcar_mmc_dma_rx_tx_data(const struct device *dev,
 	struct sdhc_data *data,
 	bool is_read)
 {
+	uintptr_t dma_addr;
+	uint32_t reg;
 	int ret = 0;
+	uint32_t dma_info1_poll_flag;
+
+	reg = rcar_mmc_read_reg32(dev, RCAR_MMC_DMA_MODE);
+	if (is_read) {
+		ret = sys_cache_data_invd_range(data->data,
+						data->blocks * data->block_size);
+		if (ret < 0) {
+			LOG_ERR("%s: can't invalidate data cache before read", dev->name);
+			return ret;
+		}
+		dma_info1_poll_flag = RCAR_MMC_DMA_INFO1_END_RD2;
+		reg |= RCAR_MMC_DMA_MODE_DIR_RD;
+	} else {
+		ret = sys_cache_data_flush_range(data->data, data->blocks * data->block_size);
+		if (ret < 0) {
+			LOG_ERR("%s: can't invalidate data cache before write",
+				dev->name);
+			return ret;
+		}
+		dma_info1_poll_flag = RCAR_MMC_DMA_INFO1_END_WR;
+		reg &= ~RCAR_MMC_DMA_MODE_DIR_RD;
+	}
+	rcar_mmc_write_reg32(dev, RCAR_MMC_DMA_MODE, reg);
+
+	reg = rcar_mmc_read_reg32(dev, RCAR_MMC_EXTMODE);
+	reg |= RCAR_MMC_EXTMODE_DMA_EN;
+	rcar_mmc_write_reg32(dev, RCAR_MMC_EXTMODE, reg);
+
+	dma_addr = z_mem_phys_addr(data->data);
+	rcar_mmc_write_reg32(dev, RCAR_MMC_DMA_ADDR_L, dma_addr);
+	rcar_mmc_write_reg32(dev, RCAR_MMC_DMA_ADDR_H, 0);
+
+	rcar_mmc_write_reg32(dev, RCAR_MMC_DMA_CTL, RCAR_MMC_DMA_CTL_START);
+
+	ret = rcar_mmc_poll_reg_flags_check_err(dev, RCAR_MMC_DMA_INFO1, dma_info1_poll_flag,
+						dma_info1_poll_flag, false, true);
+
+	if (is_read) {
+		if (sys_cache_data_invd_range(data->data, data->blocks * data->block_size) < 0) {
```

Comment:
I found info that I need an invalidation here, also the Linux Renesas MMC driver do the same
```
Invalidate: Before cpu tries to read a portion of memory updated by a device, the corresponding memory needs to be invalidated.
```

could you please share with me where did you find this info?
> All device drivers assume that no one touches data buffers when drivers are working with them.

Reply:
>  Linux Renesas MMC driver do the same

I bet it does not invalidate cache before DMA operation. Which makes sense.

On other hand, I don't see sense in invalidating cache twice.  You already invalidated cache lines for this buffer before the DMA operation start. How you can get cache entries for this buffer if CPU does not touch it?  

> could you please share with me where did you find this info?

This is sort of general knowledge. Reasoning is following: user code does not know in what exact moment driver (or hardware in case of DMA) will read/write from/to a buffer. There is no synchronization primitives for such cases. So, when user code provides buffer to a driver, it is expected that user code will not touch the buffer until driver return an operation result. 

If you want some more formal source, I can propose the following quote from Linux Device Drivers:

> Once a buffer has been mapped, it belongs to the device, not the processor. Until
the buffer has been unmapped, the driver should not touch its contents in any
way. Only after dma_unmap_single has been called is it safe for the driver to
access the contents of the buffer (with one exception that we see shortly).
Among other things, this rule implies that a buffer being written to a device cannot be mapped until it contains all the data to write. 

This is about DMA operations, but you got the idea.

Another point: `dma-buf` interface in Linux provides to clear functions: `begin_cpu_access` and `end_cpu_access` to control when CPU might access shared DMA buffers.